source/develop/release-management/features/virt/dedicated-cpu.md

```diff
@@ -0,0 +1,170 @@
+migrated back.
+
+*TODO*: Tips how to identify candidates for migration?  virsh? Something in UI?
+
+
+## User Experience
+
+## Changes to Engine and Vdsm
+
+The high level overview of the workflow in Engine and Vdsm is following:
+
+1.  When host switches to *UP* state engine queries it to find out which pCPUs
+    are dedicated through a policy or pinned. This is necessary for engine to
+    get a good picture about CPU usage so that it can schedule new VMs
+    properly.
+
+2.  On VM start/migration/etc. Engine consults the policy for dedicating CPUs
+    and schedules the VM on host with enough free CPUs. Engine determines the
+    mapping between physical and virtual CPUs.
+
+3.  Engine produces domain XML that contains properly populated `<vcpupin>`
+    element (similarly to CPU pinning) and stores the policy name in VM
+    metadata.
+
+4.  Vdsm starts the VM and analyzes which CPUs were dedicated. Vdsm updates the
+    pool of shared CPUs and re-configures VMs without any CPU pinning or
+    dedicated CPUs to run on the new set of CPUs from shared pool.
+
+5.  When the VM with CPU pinning or dedicated CPUs is being destroyed, Vdsm
+    updates the pool of shared CPUs. Vdsm then re-configures VMs without any
+    CPU pinning or dedicated CPUs to run on the new set of CPUs from shared
+    pool.
+
+Several changes to the internal API are required. To accomplish step 1. new API
+verb `Host.cpuInfo` needs to be defined that will provide:
+
+  * list of pCPUs dedicated to VMs through CPU pinning or policy
+  * CPU indices and topology information how are sockets, cores and threads
+    organized
+
+Further, Engine will add a new entry into VM metadata called `cpuPolicy` that
+will contain the requested CPU policy. And finally, migration parameters in
+`VM.migrate` will contain a field `cpus` that will contain the CPU mapping for
+the VM on the destination host. This is necessary so that the source VDSM can
+provide a correct domain XML to the destination host.
+
+*TODO*: In what format? Can we use `<vcpupin>` element content or should we use
+something more concise? Alternatively, we can also let destination VDSM parse
+`<vcpupin>` from XML in `VM.migrationCreate` and use libvirt API to pin the CPUs.
+
+## Testing
+
+*TODO*
+
+There are many parts that need to be tested 
+
+feature testing:
+- create and especially migration
+- isolate-threads is respected
+- require-threads excludes unused threads
+
+backward compatibility:
+- cpu pinning
+- migration
+
+
+## Future Work
+
+In the beginning CPU hot-plugging will not be supported for VMs with policies
+other than *shared*. To figure out how the hot-plugging should behave for each
+particular policy will remain as a future work .
+
+Use of VMs with dedicated CPU policies may in time lead to fragmentation of
+resources on the hosts. Leading not only to inefficient use, difficulties with
+scheduling but ultimately could prevent new VMs from starting on a host.
```

Comment:
It would be worth to elaborate a bit more. What kinds of problems, not already experienced without dedicated CPUs, can arise? Do you have NUMA on mind and/or something else?

Reply:
I am afraid this is little bit hard to explain shortly and without pictures. :) But yes, one such example is when you request a whole NUMA node and you cannot because even though the host may appear to have enough free pCPUs the NUMA node(s) are already occupied by smaller VMs.

Reply:
At least one example explained briefly and without pictures :-), let's use it for clarification (I'm not sure whether there can be other kinds of fragmentation).

Reply:
Updated with an example.